client-side-virtual-keyboard.md

```diff
@@ -24,87 +30,56 @@ After importing the AAR file, use the AnboxWebView in an Android project:
-           ...
-           mWebview = (AnboxWebView) findViewById(R.id.webview);
-           mWebView.addJavascriptInterface(new AppInterface(this), "AppInterface");
-           ...
-       }
-   }
-   ```
-
-3. For the application running in the Android container, whenever a text editor gains or loses focus during streaming, an IME-specific message is sent from the Android container to the client through the JavaScript SDK. Each message requests a certain IME operation. This allows you to perform one certain action when the message is received and ensures the behaviour of the virtual keyboard is synced on both ends. The supported IME messages are as follows:
-
-   The following IME messages are supported:
+## 3. Build communication bridge
 
-    Message          |  Description
-    -----------------|-----------------------------------------------
-    show-ime         |  Request to show client-side virtual keyboard
-    hide-ime         |  Request to hide client-side virtual keyboard
+In your web application code, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK.
 
-    To receive the above messages, register the `messageReceived` callback function when instantiating an AnboxStream object in HTML. Through the injected Java object `AppInterface`, the client-side virtual keyboard will pop up or hide when the corresponding message is received.
+Adding the following JavaScript code snippet builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer. It attaches the AnboxStream object to the global `window` object. The functions will automatically be invoked by the Android Java layer (AnboxWebView).
 
    ```
-       window.stream = new AnboxStream({
-           ...
-           ...
-           ...
-           callbacks: {
-               ...
-               ...
-               messageReceived: (type, data) => {
-                   switch(type) {
-                   case "show-ime" :
-                       AppInterface.openVirtualKeyboard();
-                   break;
-                   case "hide-ime" :
-                       AppInterface.hideVirtualKeyboard();
-                   break;
-                   default:
-                       console.log("Unhandled message:" + type)
-                   break;
-                   }
-               }
-           }
-       });
-   ```
+   <script type="module">
+     import {AnboxStream} from './js/anbox-stream-sdk.js';
+     window.stream = new AnboxStream({
+     ...
+     ...
+     ...
+     });
+   </script>
 
-4. Set up the listener to the activity where you want to capture the message. You must implement the following methods for the AppInterface.ActionListener interfaces so that the application will be notified when a request is coming from the JavaScript SDK.
+   <script>
+     function sendIMECommittedText(text) {
```

Comment:
Nitpick: it would be nice to say where those function should be defined, e.g.:

> If you define those functions somewhere else, make sure they are stored under the global `window` object, e.g.: running `window.sendIMECommittedText()` should work

Reply:
That's fine.  However as we've shown how to initialize the stream object and attach the stream object to the global window object, I'd prefer to keep it as is and not go further down to the actual JS related knowledge since the section is mainly about the client-side virtual keyboard integration.

Reply:
Fair enough, LGTM then